cupy/statistics/histogram.py

```diff
@@ -217,29 +217,21 @@ def histogram(x, bins=10, range=None, weights=None, density=False):
                     and x.size <= 0x7fffffff and bin_edges.size <= 0x7fffffff):
                 # Need to ensure the dtype of bin_edges as it's needed for both
                 # the CUB call and the correction later
-                if isinstance(bins, cupy.ndarray):
-                    bin_type = cupy.result_type(bin_edges, x)
-                    if cupy.issubdtype(bin_type, cupy.integer):
-                        bin_type = cupy.result_type(bin_type, float)
-                    bin_edges = bin_edges.astype(bin_type, copy=False)
+                assert isinstance(bin_edges, cupy.ndarray)
+                bin_type = numpy.result_type(bin_edges.dtype, x.dtype)
+                if numpy.issubdtype(x.dtype, numpy.integer):
+                    bin_type = numpy.float
```

Comment:
This is cleaner 👍 But, when `_get_bin_edges()` generates `bin_edges`, if `bins` is `int` this type requirement is already satisfied, so it's why we first test for `if isinstance(bins, cupy.ndarray):`. Would be nice to keep the if branch.

Reply:
Certainly, it seems to be faster.
The point is that bins needs type conversion even when `numpy.array` or `list` is an input.
How about `if not is instance(bins, int):`?

Reply:
Yeah let's prepare for #3542.